### ISSUE TITLE
False positive `no-member` error if the `__new__` static method replaces returning class

### ISSUE THREAD

**Issue by mymedia2** [PyCQA/pylint#7258]
### Bug description

The dunder method \_\_new\_\_ may return object not necessary corresponding to the first argument. Pylint could infer actual variable type after construction if astroid looked deeper into the method.

```python
from dataclasses import dataclass


class Chameleon:
    @dataclass
    class response:
        universe: int

    def __new__(cls, ultimate):
        return cls.response(ultimate * 3.5)


assert Chameleon(12).universe == 42
```

### Command used

```shell
pylint -s no -d all -e no-member chameleon1.py
```

### Pylint output

```shell
************* Module chameleon1
chameleon1.py:13:7: E1101: Instance of 'Chameleon' has no 'universe' member (no-member)
```

### Expected behavior

No false error.

### Pylint version

```shell
pylint 2.14.5
astroid 2.11.7
Python 3.10.4 (main, Jun 29 2022, 12:14:53) [GCC 11.2.0]
```

### OS / Environment

Ubuntu 22.04 (Jammy Jellyfish)

**Comment by jacobtylerwalls** [comment]
Very closely related to https://github.com/PyCQA/pylint/issues/981 and https://github.com/PyCQA/astroid/issues/457 but could potentially be solved separately, as it has nothing to do with inheritance. I appreciate your statement of the problem: pylint/astroid is doing absolutely nothing to find out what is returned by `__new__()`.
